Panic in the error handling of the OTLP ingestion in the distributor

#### Describe the bug
An invalid OTLP payload can lead to a distributor panic, because a counter can end up with a negative value.

#### To Reproduce
Unfortunately I don't know how exactly the payload looked like.

#### Expected behavior

Don't panic

#### Environment

Mimir `r203` in Grafana Cloud

#### Additional Context

This is the panic:

```
panic: counter cannot decrease in value

goroutine 249835 [running]:
github.com/prometheus/client_golang/prometheus.(*counter).Add(0xc06e174000?, 0xc0583f1a60?)
        /backend-enterprise/vendor/github.com/prometheus/client_golang/prometheus/counter.go:109 +0xf4
github.com/grafana/mimir/pkg/util/push.otelMetricsToTimeseries({0x2d27330, 0xc07bb21e30}, {0x2d08800, 0xc05922bae0}, {0x10?})
        /backend-enterprise/vendor/github.com/grafana/mimir/pkg/util/push/otel.go:131 +0x186
github.com/grafana/mimir/pkg/util/push.OTLPHandler.func1({0x2d27330, 0xc07bb21e30}, 0xc0583d9800, 0x6400000, {0xc07bb21c80?, 0xc0552a2620?, 0x2603680?}, 0xc05922d1c0)
        /backend-enterprise/vendor/github.com/grafana/mimir/pkg/util/push/otel.go:111 +0x373
github.com/grafana/mimir/pkg/util/push.handler.func1({0x2d247b0, 0xc05922d180}, 0xc0583d9800)
        /backend-enterprise/vendor/github.com/grafana/mimir/pkg/util/push/push.go:90 +0x2d1
net/http.HandlerFunc.ServeHTTP(0xc0583d9700?, {0x2d247b0?, 0xc05922d180?}, 0x2cf9c10?)
        /usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/grafana/backend-enterprise/pkg/enterprise/mimir/labelaccess.(*labelAccessMiddleware).Wrap.func1({0x2d247b0, 0xc05922d180}, 0xc0583d9700)
        /backend-enterprise/pkg/enterprise/mimir/labelaccess/middleware.go:50 +0x332
net/http.HandlerFunc.ServeHTTP(0xc0583d9600?, {0x2d247b0?, 0xc05922d180?}, 0x2cf9d01?)
        /usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/weaveworks/common/middleware.glob..func1.1({0x2d247b0, 0xc05922d180}, 0xc0583d9600)
        /backend-enterprise/vendor/github.com/weaveworks/common/middleware/http_auth.go:17 +0x110
net/http.HandlerFunc.ServeHTTP(0xc0583d9500?, {0x2d247b0?, 0xc05922d180?}, 0x20301e?)
        /usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/gorilla/mux.(*Router).ServeHTTP(0xc0004cc600, {0x2d247b0, 0xc05922d180}, 0xc0583d9400)
        /backend-enterprise/vendor/github.com/gorilla/mux/mux.go:210 +0x1cf
github.com/weaveworks/common/httpgrpc/server.Server.Handle({{0x2d08ce0?, 0xc0004cc600?}}, {0x2d27330?, 0xc07bb21c20}, 0xc05922b810)
        /backend-enterprise/vendor/github.com/weaveworks/common/httpgrpc/server/server.go:61 +0x3c3
github.com/weaveworks/common/httpgrpc._HTTP_Handle_Handler.func1({0x2d27330, 0xc07bb21c20}, {0x2650840?, 0xc05922b810})
        /backend-enterprise/vendor/github.com/weaveworks/common/httpgrpc/httpgrpc.pb.go:456 +0x78
github.com/grafana/mimir/pkg/mimir.ThanosTracerUnaryInterceptor({0x2d27330?, 0xc07bb21bf0?}, {0x2650840, 0xc05922b810}, 0xc0594389c0?, 0xc05943cb28)
        /backend-enterprise/vendor/github.com/grafana/mimir/pkg/mimir/tracing.go:25 +0x7a
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1({0x2d27330?, 0xc07bb21bf0?}, {0x2650840?, 0xc05922b810?})
        /backend-enterprise/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x3a
github.com/weaveworks/common/middleware.ServerUserHeaderInterceptor({0x2d27330?, 0xc07bb21b90?}, {0x2650840, 0xc05922b810}, 0xc05c4b15e8?, 0xc0583f17c0)
        /backend-enterprise/vendor/github.com/weaveworks/common/middleware/grpc_auth.go:38 +0x65
github.com/grafana/mimir/pkg/util/noauth.SetupAuthMiddleware.func1({0x2d27330, 0xc07bb21b90}, {0x2650840, 0xc05922b810}, 0xc0583f17a0, 0xc0583f17c0)
        /backend-enterprise/vendor/github.com/grafana/mimir/pkg/util/noauth/no_auth.go:32 +0xa7
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1({0x2d27330?, 0xc07bb21b90?}, {0x2650840?, 0xc05922b810?})
        /backend-enterprise/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x3a
github.com/weaveworks/common/middleware.UnaryServerInstrumentInterceptor.func1({0x2d27330, 0xc07bb21b90}, {0x2650840, 0xc05922b810}, 0xc0583f17a0, 0xc0583f17e0)
        /backend-enterprise/vendor/github.com/weaveworks/common/middleware/grpc_instrumentation.go:35 +0xa2
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1({0x2d27330?, 0xc07bb21b90?}, {0x2650840?, 0xc05922b810?})
        /backend-enterprise/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x3a
github.com/opentracing-contrib/go-grpc.OpenTracingServerInterceptor.func1({0x2d27330, 0xc07bb21890}, {0x2650840, 0xc05922b810}, 0xc0583f17a0, 0xc0583f1800)
        /backend-enterprise/vendor/github.com/opentracing-contrib/go-grpc/server.go:57 +0x402
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1({0x2d27330?, 0xc07bb21890?}, {0x2650840?, 0xc05922b810?})
        /backend-enterprise/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x3a
github.com/weaveworks/common/middleware.GRPCServerLog.UnaryServerInterceptor({{0x2d42b28?, 0xc000813f10?}, 0x0?, 0x40?}, {0x2d27330, 0xc07bb21890}, {0x2650840, 0xc05922b810}, 0xc0583f17a0, 0xc0583f1820)
        /backend-enterprise/vendor/github.com/weaveworks/common/middleware/grpc_logging.go:30 +0xbb
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1({0x2d27330?, 0xc07bb21890?}, {0x2650840?, 0xc05922b810?})
        /backend-enterprise/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x3a
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1({0x2d27330, 0xc07bb21890}, {0x2650840, 0xc05922b810}, 0xc048ac1b00?, 0x233d900?)
        /backend-enterprise/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:34 +0xbe
github.com/weaveworks/common/httpgrpc._HTTP_Handle_Handler({0x22ecd60?, 0xc0005da0c0}, {0x2d27330, 0xc07bb21890}, 0xc0583f7aa0, 0xc041c4bbf0)
        /backend-enterprise/vendor/github.com/weaveworks/common/httpgrpc/httpgrpc.pb.go:458 +0x138
google.golang.org/grpc.(*Server).processUnaryRPC(0xc04232e000, {0x2d41720, 0xc05a1776c0}, 0xc0594586c0, 0xc0005d2210, 0x3f30360, 0x0)
        /backend-enterprise/vendor/google.golang.org/grpc/server.go:1282 +0xccf
google.golang.org/grpc.(*Server).handleStream(0xc04232e000, {0x2d41720, 0xc05a1776c0}, 0xc0594586c0, 0x0)
        /backend-enterprise/vendor/google.golang.org/grpc/server.go:1619 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        /backend-enterprise/vendor/google.golang.org/grpc/server.go:921 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /backend-enterprise/vendor/google.golang.org/grpc/server.go:919 +0x28a
```

I temporarily patched it by commenting this counter which can end up with a negative value:

https://github.com/grafana/mimir/compare/comment_out_panicking_metric?expand=1

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -29,6 +29,7 @@
 * [BUGFIX] Querier: re-balance the querier worker connections when a query-frontend or query-scheduler is terminated. #3005
 * [BUGFIX] Distributor: Now returns the quorum error from ingesters. For example, with replication_factor=3, two HTTP 400 errors and one HTTP 500 error, now the distributor will always return HTTP 400. Previously the behaviour was to return the error which the distributor first received. #2979
 * [BUGFIX] Query-frontend: query sharding took exponential time to map binary expressions. #3027
+* [BUGFIX] Distributor: Stop panics on OTLP endpoint when a single metric has multiple timeseries. #3040
 
 ### Mixin
 

--- a/pkg/util/push/otel.go
+++ b/pkg/util/push/otel.go
@@ -14,6 +14,7 @@ import (
 	"github.com/go-kit/log/level"
 	"github.com/grafana/dskit/tenant"
 	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/prometheusremotewrite"
+	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/prompb"
 	"github.com/weaveworks/common/httpgrpc"
 	"github.com/weaveworks/common/middleware"
@@ -201,7 +202,7 @@ func TimeseriesToOTLPRequest(timeseries []prompb.TimeSeries) pmetricotlp.Request
 		attributes := pcommon.NewMap()
 
 		for _, l := range ts.Labels {
-			if l.Name == "__name__" {
+			if l.Name == model.MetricNameLabel {
 				name = l.Value
 				continue
 			}

--- a/pkg/util/push/otel.go
+++ b/pkg/util/push/otel.go
@@ -127,7 +127,7 @@ func otelMetricsToTimeseries(ctx context.Context, logger kitlog.Logger, md pmetr
 			return nil, err
 		}
 
-		dropped := md.MetricCount() - len(tsMap)
+		dropped := md.DataPointCount() - sampleCountInMap(tsMap)
 		validation.DiscardedSamples.WithLabelValues(otelParseError, userID).Add(float64(dropped))
 
 		parseErrs := errs.Error()
@@ -222,4 +222,13 @@ func TimeseriesToOTLPRequest(timeseries []prompb.TimeSeries) pmetricotlp.Request
 	}
 
 	return pmetricotlp.NewRequestFromMetrics(d)
+}
+
+func sampleCountInMap(tsMap map[string]*prompb.TimeSeries) int {
+	count := 0
+	for _, ts := range tsMap {
+		count += len(ts.Samples)
+	}
+
+	return count
 }

--- a/pkg/util/push/push_test.go
+++ b/pkg/util/push/push_test.go
@@ -18,6 +18,7 @@ import (
 
 	"github.com/gogo/protobuf/proto"
 	"github.com/golang/snappy"
+	"github.com/grafana/dskit/tenant"
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/prompb"
 	"github.com/stretchr/testify/assert"
@@ -27,7 +28,6 @@ import (
 	"go.opentelemetry.io/collector/pdata/pmetric"
 	"go.opentelemetry.io/collector/pdata/pmetric/pmetricotlp"
 
-	"github.com/grafana/dskit/tenant"
 	"github.com/grafana/mimir/pkg/mimirpb"
 )
 

--- a/pkg/util/push/push_test.go
+++ b/pkg/util/push/push_test.go
@@ -18,12 +18,16 @@ import (
 
 	"github.com/gogo/protobuf/proto"
 	"github.com/golang/snappy"
+	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/prompb"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"github.com/weaveworks/common/middleware"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
 	"go.opentelemetry.io/collector/pdata/pmetric/pmetricotlp"
 
+	"github.com/grafana/dskit/tenant"
 	"github.com/grafana/mimir/pkg/mimirpb"
 )
 
@@ -39,6 +43,56 @@ func TestHandler_otlpWriteNoCompression(t *testing.T) {
 	req := createOTLPRequest(t, createOTLPMetricRequest(t), false)
 	resp := httptest.NewRecorder()
 	handler := OTLPHandler(100000, nil, false, verifyWriteRequestHandler(t, mimirpb.API))
+	handler.ServeHTTP(resp, req)
+	assert.Equal(t, 200, resp.Code)
+}
+
+func TestHandler_otlpDroppedMetricsPanic(t *testing.T) {
+	// https://github.com/grafana/mimir/issues/3037 is triggered by a single metric
+	// having two different datapoints that correspond to different Prometheus metrics.
+
+	// For the error to be triggered, md.MetricCount() < len(tsMap), hence we're inserting 3 valid
+	// samples from one metric (len = 3), and one invalid metric (metric count = 2).
+
+	md := pmetric.NewMetrics()
+	name := "foo"
+	attributes := pcommon.NewMap()
+	attributes.InsertString(model.MetricNameLabel, name)
+
+	metric1 := md.ResourceMetrics().AppendEmpty().ScopeMetrics().AppendEmpty().Metrics().AppendEmpty()
+	metric1.SetName(name)
+	metric1.SetDataType(pmetric.MetricDataTypeGauge)
+
+	datapoint1 := metric1.Gauge().DataPoints().AppendEmpty()
+	datapoint1.SetTimestamp(pcommon.NewTimestampFromTime(time.Now()))
+	datapoint1.SetDoubleVal(0)
+	attributes.CopyTo(datapoint1.Attributes())
+	datapoint1.Attributes().InsertString("diff_label", "bar")
+
+	datapoint2 := metric1.Gauge().DataPoints().AppendEmpty()
+	datapoint2.SetTimestamp(pcommon.NewTimestampFromTime(time.Now()))
+	datapoint2.SetDoubleVal(0)
+	attributes.CopyTo(datapoint2.Attributes())
+	datapoint2.Attributes().InsertString("diff_label", "baz")
+
+	datapoint3 := metric1.Gauge().DataPoints().AppendEmpty()
+	datapoint3.SetTimestamp(pcommon.NewTimestampFromTime(time.Now()))
+	datapoint3.SetDoubleVal(0)
+	attributes.CopyTo(datapoint3.Attributes())
+	datapoint3.Attributes().InsertString("diff_label", "food")
+
+	metric2 := md.ResourceMetrics().AppendEmpty().ScopeMetrics().AppendEmpty().Metrics().AppendEmpty()
+	metric2.SetName(name)
+	metric2.SetDataType(pmetric.MetricDataTypeGauge)
+
+	req := createOTLPRequest(t, pmetricotlp.NewRequestFromMetrics(md), false)
+	resp := httptest.NewRecorder()
+	handler := OTLPHandler(100000, nil, false, func(ctx context.Context, request *mimirpb.WriteRequest, cleanup func()) (response *mimirpb.WriteResponse, err error) {
+		assert.Len(t, request.Timeseries, 3)
+		assert.False(t, request.SkipLabelNameValidation)
+		cleanup()
+		return &mimirpb.WriteResponse{}, nil
+	})
 	handler.ServeHTTP(resp, req)
 	assert.Equal(t, 200, resp.Code)
 }
@@ -264,6 +318,13 @@ func createOTLPRequest(t testing.TB, metricRequest pmetricotlp.Request, compress
 	req, err := http.NewRequest("POST", "http://localhost/", bytes.NewReader(body))
 	require.NoError(t, err)
 	req.Header.Set("Content-Type", "application/x-protobuf")
+	req.Header.Set("X-Scope-OrgID", "test")
+
+	// We need this for testing dropped metrics codepath which requires
+	// tenantID to be present.
+	_, ctx, err := tenant.ExtractTenantIDFromHTTPRequest(req)
+	require.NoError(t, err)
+	req = req.WithContext(ctx)
 
 	if compress {
 		req.Header.Set("Content-Encoding", "gzip")

--- a/pkg/util/push/push_test.go
+++ b/pkg/util/push/push_test.go
@@ -55,7 +55,7 @@ func TestHandler_otlpDroppedMetricsPanic(t *testing.T) {
 	// samples from one metric (len = 3), and one invalid metric (metric count = 2).
 
 	md := pmetric.NewMetrics()
-	name := "foo"
+	const name = "foo"
 	attributes := pcommon.NewMap()
 	attributes.InsertString(model.MetricNameLabel, name)
 

